Presearch News package. Embedded image

piwnik opened the pull request:
* Embedding videos thumbnail images to response
* Added "more videos" button
* Reduced returned videos size from 10 to 5 (to reduce response time/weight)

Response time increases to 2.2s, response size ~150kb

![presearch more](https://user-images.githubusercontent.com/3851067/201446868-d056ed37-1fd4-45fd-aa70-06e8e47ba2d2.gif)

@jejopl @treygrainger

treygrainger commented on packages/presearchNews/index.js:
```diff
@@ -269,26 +317,42 @@ async function presearchNews(query, API_KEY) {
+
+        const items = [];
+
+        for (let i = 0; i < data.result.items.length; i++) {
```
you're not doing parallelizing this, so if it's taking 2.2 seconds like you said, and you're loading 6 images serially, then that's 366ms on average per image.

If you do them in parallel, then you'll be under the allowed latency threshold.

treygrainger replied:
Maybe just create a list of promises for each of the urlToBase64(image) lookups and then do something along the lines of:

Promises.all(data.results.items.map(url => urlToBase64(url)).then((responses) => { //process the array of parallel base64 results }).catch((error) = { ... } );

That way, all the images can be fetched and converted to base64 in parallel instead of sequentially.

treygrainger replied:
(note: I didn't test that, so don't expects it to work first try... you'll probably need to adjust; was just trying to provide a general idea).

piwnik replied:
Yes, it definitely improves performance, thank you (just got same idea when woke up)! After speed measurements I figured out, that now bigger problem is video search on lbry, which takes ~800ms. I'll investigate video fetch alternatives.

piwnik commented:
1. Performance. 
1.1. Changed video source source from LBRY to odysee proxy, this improves videos info fetch twice.
1.2. Overall extension renders in range of 500ms-800ms. Sometimes randomly image load can took up to 2s. This will be solved with cache implementation.

2. UI changes, changed a bit style for link, now it goes blue on hover.

3. Articles
3.1. Medium API is more designed for maintain "your own" articles, than for fetch some other channel/user. I's possible to fetch articles via Medium API (https://github.com/Medium/medium-api-docs), but then integration token and presearch userid is required for you, because no way to get someone else userid than your self.
3.2. Medium have rss feed (https://medium.com/feed/@presearch), which not requires any token and publicly available for any channel/user articles, but problem is that it returns data only in xml format, so conversion need to be done. 
3.2.1. It can be done by additionally installed 3-rd party package like https://www.npmjs.com/package/rss-parser
3.2.2. It can be done by requesting feed via proxy like  https://api.rss2json.com/v1/api.json?rss_url=https://medium.com/feed/@presearch
3.2.3. Write local parser inside package.
3.3. So need decide, what API package should use

3.4. I have few example how articles form medium can fit, so need to choose
1. Same way like videos
![image](https://user-images.githubusercontent.com/3851067/201530463-cb04cd91-750a-4913-8fda-2bf6d37c0085.png)
2. Same way like videos but without thumbnails
![image](https://user-images.githubusercontent.com/3851067/201530497-cd2652bc-cbcf-4b06-926b-1ae343442614.png)
3. Show list with 3 (or more if decided) items, where to read more medium page will be opened
![image](https://user-images.githubusercontent.com/3851067/201530536-ce356a17-c908-40d7-9c38-f441f1a7bf2a.png)
4. Show list with 3 (or more if decided) items, with expand and view 7 more possibility 
![image](https://user-images.githubusercontent.com/3851067/201530612-95ede69e-4525-4bd0-ac09-adeea6bb3355.png)

@treygrainger @jejopl would be good know your opinion, if we should show articles at all, if yes - which approach use to fetch them, what variant to choose for UI.

treygrainger commented:
@piwnik Awesome to see the progress. Personally, I'd prefer to see a flat list with the videos and medium articles merged. So basically using the thumbnails for Medium (like #1), but having a flat list. If the most recent news item was a medium post, that shows up, and if the most recent was a video, that shows up.

There's no guarantee that we always do a weekly video update forever or whether the medium articles will be more recent vs. the videos.

You could visually overlay a "Play" icon over the videos to distinguish video content from articles.

Leveraging this format, we could easily add in Tweets or other future content, as well, without needing to change the layout.

WDYT?

![image](https://user-images.githubusercontent.com/1043085/201532446-d94dea6d-6c93-4a7f-ac47-4708b071fd4d.png)

piwnik commented:
Agree, good idea, will be implemented.

piwnik commented:
![presearch news](https://user-images.githubusercontent.com/3851067/201793704-29c1e2d4-20f9-4d22-a32a-39f1d4cf5162.gif)

* Medium articles support
* Simplified UI (removed "more" links & expand description)
* Added "play" overlay for videos

@treygrainger what do you think?
